go-controller/pkg/clustermanager/networkconnect/cluster_network_connect.go

```diff
@@ -0,0 +1,319 @@
+			}
+		default:
+			return nil, nil, fmt.Errorf("%w: unsupported network selection type %s", errConfig, selector.NetworkSelectionType)
+		}
+	}
+
+	return discoveredNetworks, allMatchingNADKeys, nil
+}
+
+func computeNetworkOwner(networkType string, networkID int) string {
+	return fmt.Sprintf("%s_%d", networkType, networkID)
+}
+
+// parseNetworkOwner extracts the topology type from an owner key.
+// Owner keys are formatted as "{topology}_{networkID}" (e.g., "layer3_1", "layer2_2").
+func parseNetworkOwner(owner string) (topologyType string, ok bool) {
+	if len(owner) > len(ovntypes.Layer3Topology)+1 && owner[:len(ovntypes.Layer3Topology)] == ovntypes.Layer3Topology {
+		return ovntypes.Layer3Topology, true
+	}
+	if len(owner) > len(ovntypes.Layer2Topology)+1 && owner[:len(ovntypes.Layer2Topology)] == ovntypes.Layer2Topology {
+		return ovntypes.Layer2Topology, true
+	}
+	return "", false
+}
+
+// allocateSubnets allocates subnets for the given discovered networks
+// It returns a map of owner to subnets
+// NOTE: If owner already had its subnets allocated, it will simply return those existing subnets
+func (c *Controller) allocateSubnets(discoveredNetworks []*util.NetInfo, connectSubnets []networkconnectv1.ConnectSubnet,
+	cncName string) (map[string][]*net.IPNet, sets.Set[string], error) {
+	connectSubnetAllocator, err := c.fetchConnectSubnetAllocator(connectSubnets, cncName)
+	if err != nil {
+		return nil, nil, err
+	}
+	var owner string
+	var subnets []*net.IPNet
+	allMatchingNetworkKeys := sets.New[string]()
+	allocatedSubnets := make(map[string][]*net.IPNet)
+	for _, network := range discoveredNetworks {
+		netInfo := *network
+		networkID := netInfo.GetNetworkID()
+		if networkID == ovntypes.NoNetworkID {
+			return nil, nil, fmt.Errorf("network id is invalid for network %s", netInfo.GetNetworkName())
+		}
+		if netInfo.TopologyType() == ovntypes.Layer3Topology {
+			owner = computeNetworkOwner(ovntypes.Layer3Topology, networkID)
+			subnets, err = connectSubnetAllocator.AllocateLayer3Subnet(owner)
+			if err != nil {
+				return nil, nil, err
+			}
+		} else if netInfo.TopologyType() == ovntypes.Layer2Topology {
+			owner = computeNetworkOwner(ovntypes.Layer2Topology, networkID)
+			subnets, err = connectSubnetAllocator.AllocateLayer2Subnet(owner)
+			if err != nil {
+				return nil, nil, err
+			}
+		} else {
+			// This should never happen
+			return nil, nil, fmt.Errorf("unsupported network topology type %s for network %s", netInfo.TopologyType(), netInfo.GetNetworkName())
+		}
+		allocatedSubnets[owner] = subnets
+		allMatchingNetworkKeys.Insert(owner)
+		klog.V(5).Infof("Allocated subnets %v for %s (network: %s)", subnets, owner, netInfo.GetNetworkName())
+	}
+	return allocatedSubnets, allMatchingNetworkKeys, nil
+}
+
+func (c *Controller) fetchConnectSubnetAllocator(connectSubnets []networkconnectv1.ConnectSubnet, cncName string) (HybridConnectSubnetAllocator, error) {
```

Comment:
I think it would be easier to create and initialize SubnetAllocator on the `clusterNetworkConnectState` creation and then afterwards just use it as `clusterNetworkConnectState.allocator`
Constantly passing `connectSubnets` confuses me :/

Reply:
actually I had meant to do this.. face palm.. its in my cleanup notes but realising I never got around to doing this. I 100% agree with you.

Reply:
done.